File: backend/core/permissions.py
# backend/core/permissions.py
from django.contrib.auth.models import Permission, Group
from django.contrib.contenttypes.models import ContentType
from django.contrib.auth import get_user_model

User = get_user_model()

# Definição de todas as permissões do sistema
SYSTEM_PERMISSIONS = {
    # Permissões do Sistema
    'system': [
        ('system.manage_settings', 'Can manage system settings'),
        ('system.view_logs', 'Can view system logs'),
        ('system.backup_restore', 'Can backup and restore system'),
        ('system.maintenance_mode', 'Can enable maintenance mode'),
    ],
    
    # Permissões de Usuários
    'users': [
        ('users.create', 'Can create users'),
        ('users.edit', 'Can edit users'),
        ('users.delete', 'Can delete users'),
        ('users.view_all', 'Can view all users'),
        ('users.change_permissions', 'Can change user permissions'),
        ('users.impersonate', 'Can impersonate users'),
        ('users.view_blog_team_only', 'Can view blog team only'),
        ('users.view_project_team_only', 'Can view project team only'),
        ('users.view_community_team_only', 'Can view community team only'),
        ('users.view_basic_info_only', 'Can view basic user info only'),
    ],
    
    # Permissões do Blog
    'blog': [
        ('blog.create_post', 'Can create blog posts'),
        ('blog.edit_own_post', 'Can edit own blog posts'),
        ('blog.edit_others_post', 'Can edit others blog posts'),
        ('blog.edit_any_post', 'Can edit any blog post'),
        ('blog.delete_own_post', 'Can delete own blog posts'),
        ('blog.delete_others_post', 'Can delete others blog posts'),
        ('blog.delete_any_post', 'Can delete any blog post'),
        ('blog.publish_post', 'Can publish blog posts'),
        ('blog.unpublish_post', 'Can unpublish blog posts'),
        ('blog.schedule_post', 'Can schedule blog posts'),
        ('blog.create_category', 'Can create blog categories'),
        ('blog.edit_category', 'Can edit blog categories'),
        ('blog.delete_category', 'Can delete blog categories'),
        ('blog.manage_tags', 'Can manage blog tags'),
        ('blog.upload_media', 'Can upload blog media'),
        ('blog.manage_seo', 'Can manage blog SEO'),
        ('blog.view_analytics', 'Can view blog analytics'),
        ('blog.moderate_comments', 'Can moderate blog comments'),
        ('blog.approve_comments', 'Can approve blog comments'),
        ('blog.delete_comments', 'Can delete blog comments'),
        ('blog.view_posts', 'Can view blog posts'),
        ('blog.view_categories', 'Can view blog categories'),
        ('blog.view_comments', 'Can view blog comments'),
        ('blog.view_basic_analytics', 'Can view basic blog analytics'),
        ('blog.view_all_posts', 'Can view all blog posts'),
        ('blog.view_own_posts', 'Can view own blog posts'),
    ],
    
    # Permissões de Projetos
    'projects': [
        ('projects.create', 'Can create projects'),
        ('projects.edit_any', 'Can edit any project'),
        ('projects.edit_own', 'Can edit own projects'),
        ('projects.delete_any', 'Can delete any project'),
        ('projects.view_all', 'Can view all projects'),
        ('projects.view_assigned', 'Can view assigned projects'),
        ('projects.approve', 'Can approve projects'),
        ('projects.approve_own', 'Can approve own projects'),
        ('projects.close_project', 'Can close projects'),
        ('projects.archive_project', 'Can archive projects'),
        ('projects.manage_budget', 'Can manage project budget'),
        ('projects.allocate_resources', 'Can allocate project resources'),
        ('projects.track_expenses', 'Can track project expenses'),
        ('projects.approve_expenses', 'Can approve project expenses'),
        ('projects.assign_volunteers', 'Can assign volunteers to projects'),
        ('projects.manage_team', 'Can manage project team'),
        ('projects.view_volunteer_profiles', 'Can view volunteer profiles'),
        ('projects.link_beneficiaries', 'Can link beneficiaries to projects'),
        ('projects.view_project_beneficiaries', 'Can view project beneficiaries'),
        ('projects.generate_reports', 'Can generate project reports'),
        ('projects.view_analytics', 'Can view project analytics'),
        ('projects.export_project_data', 'Can export project data'),
        ('projects.manage_project_partnerships', 'Can manage project partnerships'),
        ('projects.view_basic_reports', 'Can view basic project reports'),
        ('projects.view_team_assignments', 'Can view team assignments'),
        ('projects.view_basic_budget', 'Can view basic budget info'),
    ],
    
    # Permissões da Comunidade
    'community': [
        ('community.view_volunteer_applications', 'Can view volunteer applications'),
        ('community.approve_volunteers', 'Can approve volunteers'),
        ('community.reject_volunteers', 'Can reject volunteers'),
        ('community.edit_volunteer_profiles', 'Can edit volunteer profiles'),
        ('community.deactivate_volunteers', 'Can deactivate volunteers'),
        ('community.assign_volunteer_skills', 'Can assign volunteer skills'),
        ('community.manage_volunteer_training', 'Can manage volunteer training'),
        ('community.register_beneficiaries', 'Can register beneficiaries'),
        ('community.edit_beneficiary_profiles', 'Can edit beneficiary profiles'),
        ('community.approve_beneficiaries', 'Can approve beneficiaries'),
        ('community.view_beneficiary_data', 'Can view beneficiary data'),
        ('community.generate_impact_reports', 'Can generate impact reports'),
        ('community.create_partnerships', 'Can create partnerships'),
        ('community.edit_partnerships', 'Can edit partnerships'),
        ('community.approve_partnerships', 'Can approve partnerships'),
        ('community.manage_partnership_agreements', 'Can manage partnership agreements'),
        ('community.view_donor_profiles', 'Can view donor profiles'),
        ('community.manage_donor_relationships', 'Can manage donor relationships'),
        ('community.process_donations', 'Can process donations'),
        ('community.generate_donor_reports', 'Can generate donor reports'),
        ('community.send_donor_communications', 'Can send donor communications'),
        ('community.send_newsletters', 'Can send newsletters'),
        ('community.manage_communications', 'Can manage communications'),
        ('community.generate_community_reports', 'Can generate community reports'),
        ('community.export_community_data', 'Can export community data'),
        ('community.view_volunteer_list', 'Can view volunteer list'),
        ('community.view_beneficiary_list', 'Can view beneficiary list'),
        ('community.view_partnership_list', 'Can view partnership list'),
        ('community.view_donor_list', 'Can view donor list'),
        ('community.view_basic_community_data', 'Can view basic community data'),
        ('community.view_sensitive_data', 'Can view sensitive community data'),
    ],
    
    # Permissões de Relatórios
    'reports': [
        ('reports.generate_all', 'Can generate all reports'),
        ('reports.export_sensitive', 'Can export sensitive data'),
        ('reports.view_financial', 'Can view financial reports'),
        ('reports.view_summary_reports', 'Can view summary reports'),
        ('reports.view_public_analytics', 'Can view public analytics'),
    ],
}

# Definição dos grupos e suas permissões
GROUPS_PERMISSIONS = {
    'Super Admin': [
        # Sistema
        'system.manage_settings', 'system.view_logs', 'system.backup_restore', 'system.maintenance_mode',
        # Usuários
        'users.create', 'users.edit', 'users.delete', 'users.view_all', 'users.change_permissions', 'users.impersonate',
        # Blog
        'blog.create_post', 'blog.edit_any_post', 'blog.delete_any_post', 'blog.publish_post', 
        'blog.manage_categories', 'blog.moderate_comments', 'blog.view_analytics',
        # Projetos
        'projects.create', 'projects.edit_any', 'projects.delete_any', 'projects.approve', 
        'projects.close_project', 'projects.generate_reports', 'projects.manage_budget', 'projects.assign_team',
        # Comunidade
        'community.approve_volunteers', 'community.reject_volunteers', 'community.manage_partnerships',
        'community.approve_beneficiaries', 'community.manage_donor_relationships', 'community.view_sensitive_data',
        'community.export_community_data',
        # Relatórios
        'reports.generate_all', 'reports.export_sensitive', 'reports.view_financial',
    ],
    
    'Gestor de Blog': [
        'blog.create_post', 'blog.edit_own_post', 'blog.edit_others_post', 'blog.delete_own_post',
        'blog.delete_others_post', 'blog.publish_post', 'blog.unpublish_post', 'blog.schedule_post',
        'blog.create_category', 'blog.edit_category', 'blog.delete_category', 'blog.manage_tags',
        'blog.upload_media', 'blog.manage_seo', 'blog.view_analytics', 'blog.moderate_comments',
        'blog.approve_comments', 'blog.delete_comments', 'users.view_blog_team_only',
    ],
    
    'Gestor de Projetos': [
        'projects.create', 'projects.edit_any', 'projects.view_all', 'projects.approve_own',
        'projects.close_project', 'projects.archive_project', 'projects.manage_budget',
        'projects.allocate_resources', 'projects.track_expenses', 'projects.approve_expenses',
        'projects.assign_volunteers', 'projects.manage_team', 'projects.view_volunteer_profiles',
        'projects.link_beneficiaries', 'projects.view_project_beneficiaries', 'projects.generate_reports',
        'projects.view_analytics', 'projects.export_project_data', 'projects.manage_project_partnerships',
        'users.view_project_team_only',
    ],
    
    'Gestor de Comunidade': [
        'community.view_volunteer_applications', 'community.approve_volunteers', 'community.reject_volunteers',
        'community.edit_volunteer_profiles', 'community.deactivate_volunteers', 'community.assign_volunteer_skills',
        'community.manage_volunteer_training', 'community.register_beneficiaries', 'community.edit_beneficiary_profiles',
        'community.approve_beneficiaries', 'community.view_beneficiary_data', 'community.generate_impact_reports',
        'community.create_partnerships', 'community.edit_partnerships', 'community.approve_partnerships',
        'community.manage_partnership_agreements', 'community.view_donor_profiles', 'community.manage_donor_relationships',
        'community.process_donations', 'community.generate_donor_reports', 'community.send_donor_communications',
        'community.send_newsletters', 'community.manage_communications', 'community.generate_community_reports',
        'community.export_community_data', 'users.view_community_team_only',
    ],
    
    'Visualizador': [
        'blog.view_posts', 'blog.view_categories', 'blog.view_comments', 'blog.view_basic_analytics',
        'projects.view_all', 'projects.view_basic_reports', 'projects.view_team_assignments', 'projects.view_basic_budget',
        'community.view_volunteer_list', 'community.view_beneficiary_list', 'community.view_partnership_list',
        'community.view_donor_list', 'community.view_basic_community_data', 'reports.view_summary_reports',
        'reports.view_public_analytics', 'users.view_basic_info_only',
    ],
}

def create_custom_permissions():
    """Cria todas as permissões customizadas do sistema"""
    content_type = ContentType.objects.get_for_model(User)
    
    for module, permissions in SYSTEM_PERMISSIONS.items():
        for codename, name in permissions:
            permission, created = Permission.objects.get_or_create(
                codename=codename,
                defaults={
                    'name': name,
                    'content_type': content_type,
                }
            )
            if created:
                print(f'Permissão criada: {codename}')

def create_groups_and_assign_permissions():
    """Cria os grupos e atribui as permissões"""
    for group_name, permission_codenames in GROUPS_PERMISSIONS.items():
        group, created = Group.objects.get_or_create(name=group_name)
        
        if created:
            print(f'Grupo criado: {group_name}')
        
        # Limpar permissões existentes
        group.permissions.clear()
        
        # Adicionar novas permissões
        for codename in permission_codenames:
            try:
                permission = Permission.objects.get(codename=codename)
                group.permissions.add(permission)
            except Permission.DoesNotExist:
                print(f'Permissão não encontrada: {codename}')
        
        print(f'Permissões atribuídas ao grupo {group_name}: {len(permission_codenames)}')

def assign_user_to_group(user, group_name):
    """Atribui um usuário a um grupo específico"""
    try:
        group = Group.objects.get(name=group_name)
        user.groups.add(group)
        print(f'Usuário {user.username} adicionado ao grupo {group_name}')
        return True
    except Group.DoesNotExist:
        print(f'Grupo {group_name} não encontrado')
        return False

File: src/constants/permissions.ts
// src/constants/permissions.ts

export const PERMISSIONS = {
  // Sistema
  SYSTEM: {
    MANAGE_SETTINGS: 'system_manage_settings',
    VIEW_LOGS: 'system_view_logs',
    BACKUP_RESTORE: 'system_backup_restore',
    MAINTENANCE_MODE: 'system_maintenance_mode',
  },

  // Usuários
  USERS: {
    CREATE: 'users_create',
    EDIT: 'users_edit',
    DELETE: 'users_delete',
    VIEW_ALL: 'users_view_all',
    CHANGE_PERMISSIONS: 'users_change_permissions',
    IMPERSONATE: 'users_impersonate',
    VIEW_BLOG_TEAM_ONLY: 'users_view_blog_team_only',
    VIEW_PROJECT_TEAM_ONLY: 'users_view_project_team_only',
    VIEW_COMMUNITY_TEAM_ONLY: 'users_view_community_team_only',
    VIEW_BASIC_INFO_ONLY: 'users_view_basic_info_only',
  },

  // Blog
  BLOG: {
    CREATE_POST: 'blog_create_post',
    EDIT_OWN_POST: 'blog_edit_own_post',
    EDIT_OTHERS_POST: 'blog_edit_others_post',
    EDIT_ANY_POST: 'blog_edit_any_post',
    DELETE_OWN_POST: 'blog_delete_own_post',
    DELETE_OTHERS_POST: 'blog_delete_others_post',
    DELETE_ANY_POST: 'blog_delete_any_post',
    PUBLISH_POST: 'blog_publish_post',
    UNPUBLISH_POST: 'blog_unpublish_post',
    SCHEDULE_POST: 'blog_schedule_post',
    CREATE_CATEGORY: 'blog_create_category',
    EDIT_CATEGORY: 'blog_edit_category',
    DELETE_CATEGORY: 'blog_delete_category',
    MANAGE_TAGS: 'blog_manage_tags',
    UPLOAD_MEDIA: 'blog_upload_media',
    MANAGE_SEO: 'blog_manage_seo',
    VIEW_ANALYTICS: 'blog_view_analytics',
    MODERATE_COMMENTS: 'blog_moderate_comments',
    APPROVE_COMMENTS: 'blog_approve_comments',
    DELETE_COMMENTS: 'blog_delete_comments',
    VIEW_POSTS: 'blog_view_posts',
    VIEW_CATEGORIES: 'blog_view_categories',
    VIEW_COMMENTS: 'blog_view_comments',
    VIEW_BASIC_ANALYTICS: 'blog_view_basic_analytics',
    VIEW_ALL_POSTS: 'blog_view_all_posts',
    VIEW_OWN_POSTS: 'blog_view_own_posts',
  },

  // Projetos
  PROJECTS: {
    CREATE: 'projects_create',
    EDIT_ANY: 'projects_edit_any',
    EDIT_OWN: 'projects_edit_own',
    DELETE_ANY: 'projects_delete_any',
    VIEW_ALL: 'projects_view_all',
    VIEW_ASSIGNED: 'projects_view_assigned',
    APPROVE: 'projects_approve',
    APPROVE_OWN: 'projects_approve_own',
    CLOSE_PROJECT: 'projects_close_project',
    ARCHIVE: 'projects_archive',
    MANAGE_BUDGET: 'projects_manage_budget',
    ALLOCATE_RESOURCES: 'projects_allocate_resources',
    TRACK_EXPENSES: 'projects_track_expenses',
    APPROVE_EXPENSES: 'projects_approve_expenses',
    ASSIGN_VOLUNTEERS: 'projects_assign_volunteers',
    MANAGE_TEAM: 'projects_manage_team',
    VIEW_VOLUNTEER_PROFILES: 'projects_view_volunteer_profiles',
    LINK_BENEFICIARIES: 'projects_link_beneficiaries',
    VIEW_BENEFICIARIES: 'projects_view_beneficiaries',
    GENERATE_REPORTS: 'projects_generate_reports',
    VIEW_ANALYTICS: 'projects_view_analytics',
    EXPORT_DATA: 'projects_export_data',
    MANAGE_PARTNERSHIPS: 'projects_manage_partnerships',
    VIEW_BASIC_REPORTS: 'projects_view_basic_reports',
    VIEW_TEAM_ASSIGNMENTS: 'projects_view_team_assignments',
    VIEW_BASIC_BUDGET: 'projects_view_basic_budget',
  },

  // Comunidade
  COMMUNITY: {
    VIEW_VOLUNTEER_APPLICATIONS: 'community_view_volunteer_applications',
    APPROVE_VOLUNTEERS: 'community_approve_volunteers',
    REJECT_VOLUNTEERS: 'community_reject_volunteers',
    EDIT_VOLUNTEER_PROFILES: 'community_edit_volunteer_profiles',
    DEACTIVATE_VOLUNTEERS: 'community_deactivate_volunteers',
    ASSIGN_VOLUNTEER_SKILLS: 'community_assign_volunteer_skills',
    MANAGE_VOLUNTEER_TRAINING: 'community_manage_volunteer_training',
    REGISTER_BENEFICIARIES: 'community_register_beneficiaries',
    EDIT_BENEFICIARY_PROFILES: 'community_edit_beneficiary_profiles',
    APPROVE_BENEFICIARIES: 'community_approve_beneficiaries',
    VIEW_BENEFICIARY_DATA: 'community_view_beneficiary_data',
    GENERATE_IMPACT_REPORTS: 'community_generate_impact_reports',
    CREATE_PARTNERSHIPS: 'community_create_partnerships',
    EDIT_PARTNERSHIPS: 'community_edit_partnerships',
    APPROVE_PARTNERSHIPS: 'community_approve_partnerships',
    MANAGE_PARTNERSHIP_AGREEMENTS: 'community_manage_partnership_agreements',
    VIEW_DONOR_PROFILES: 'community_view_donor_profiles',
    MANAGE_DONOR_RELATIONSHIPS: 'community_manage_donor_relationships',
    PROCESS_DONATIONS: 'community_process_donations',
    GENERATE_DONOR_REPORTS: 'community_generate_donor_reports',
    SEND_DONOR_COMMUNICATIONS: 'community_send_donor_communications',
    SEND_NEWSLETTERS: 'community_send_newsletters',
    MANAGE_COMMUNICATIONS: 'community_manage_communications',
    GENERATE_COMMUNITY_REPORTS: 'community_generate_community_reports',
    EXPORT_COMMUNITY_DATA: 'community_export_community_data',
    VIEW_VOLUNTEER_LIST: 'community_view_volunteer_list',
    VIEW_BENEFICIARY_LIST: 'community_view_beneficiary_list',
    VIEW_PARTNERSHIP_LIST: 'community_view_partnership_list',
    VIEW_DONOR_LIST: 'community_view_donor_list',
    VIEW_BASIC_COMMUNITY_DATA: 'community_view_basic_community_data',
    VIEW_SENSITIVE_COMMUNITY_DATA: 'community_view_sensitive_community_data',
  },

  // Relatórios
  REPORTS: {
    GENERATE_ALL: 'reports_generate_all',
    EXPORT_SENSITIVE: 'reports_export_sensitive',
    VIEW_FINANCIAL: 'reports_view_financial',
    VIEW_SUMMARY: 'reports_view_summary_reports',
    VIEW_PUBLIC_ANALYTICS: 'reports_view_public_analytics',
  },
} as const;

export const ROLES = {
  SUPER_ADMIN: 'Super Admin',
  BLOG_MANAGER: 'Gestor de Blog',
  PROJECT_MANAGER: 'Gestor de Projetos',
  COMMUNITY_MANAGER: 'Gestor de Comunidade',
  VIEWER: 'Visualizador',
} as const;

// Utilitários para verificação de permissões
export const checkPermission = (userPermissions: string[], permission: string): boolean => {
  return userPermissions.includes(permission);
};

export const checkAnyPermission = (userPermissions: string[], permissions: string[]): boolean => {
  return permissions.some(permission => userPermissions.includes(permission));
};

export const checkAllPermissions = (userPermissions: string[], permissions: string[]): boolean => {
  return permissions.every(permission => userPermissions.includes(permission));
};

export const checkRole = (userGroups: string[], role: string): boolean => {
  return userGroups.includes(role);
};

// Grupos de permissões comuns
export const PERMISSION_GROUPS = {
  // Permissões de administração geral
  ADMIN: [
    PERMISSIONS.SYSTEM.MANAGE_SETTINGS,
    PERMISSIONS.SYSTEM.VIEW_LOGS,
    PERMISSIONS.USERS.VIEW_ALL,
    PERMISSIONS.USERS.CREATE,
    PERMISSIONS.USERS.EDIT,
  ],

  // Permissões básicas de blog
  BLOG_BASIC: [
    PERMISSIONS.BLOG.VIEW_POSTS,
    PERMISSIONS.BLOG.VIEW_CATEGORIES,
    PERMISSIONS.BLOG.VIEW_COMMENTS,
  ],

  // Permissões de gestão de blog
  BLOG_MANAGEMENT: [
    PERMISSIONS.BLOG.CREATE_POST,
    PERMISSIONS.BLOG.EDIT_OWN_POST,
    PERMISSIONS.BLOG.PUBLISH_POST,
    PERMISSIONS.BLOG.MANAGE_TAGS,
    PERMISSIONS.BLOG.MODERATE_COMMENTS,
  ],

  // Permissões básicas de projetos
  PROJECTS_BASIC: [
    PERMISSIONS.PROJECTS.VIEW_ALL,
    PERMISSIONS.PROJECTS.VIEW_BASIC_REPORTS,
    PERMISSIONS.PROJECTS.VIEW_TEAM_ASSIGNMENTS,
  ],

  // Permissões de gestão de projetos
  PROJECTS_MANAGEMENT: [
    PERMISSIONS.PROJECTS.CREATE,
    PERMISSIONS.PROJECTS.EDIT_OWN,
    PERMISSIONS.PROJECTS.APPROVE_OWN,
    PERMISSIONS.PROJECTS.MANAGE_BUDGET,
    PERMISSIONS.PROJECTS.ASSIGN_VOLUNTEERS,
  ],

  // Permissões básicas de comunidade
  COMMUNITY_BASIC: [
    PERMISSIONS.COMMUNITY.VIEW_VOLUNTEER_LIST,
    PERMISSIONS.COMMUNITY.VIEW_BENEFICIARY_LIST,
    PERMISSIONS.COMMUNITY.VIEW_BASIC_COMMUNITY_DATA,
  ],

  // Permissões de gestão de comunidade
  COMMUNITY_MANAGEMENT: [
    PERMISSIONS.COMMUNITY.APPROVE_VOLUNTEERS,
    PERMISSIONS.COMMUNITY.REGISTER_BENEFICIARIES,
    PERMISSIONS.COMMUNITY.CREATE_PARTNERSHIPS,
    PERMISSIONS.COMMUNITY.MANAGE_DONOR_RELATIONSHIPS,
  ],

  // Permissões básicas de relatórios
  REPORTS_BASIC: [
    PERMISSIONS.REPORTS.VIEW_SUMMARY,
    PERMISSIONS.REPORTS.VIEW_PUBLIC_ANALYTICS,
  ],

  // Permissões avançadas de relatórios
  REPORTS_ADVANCED: [
    PERMISSIONS.REPORTS.GENERATE_ALL,
    PERMISSIONS.REPORTS.VIEW_FINANCIAL,
    PERMISSIONS.REPORTS.EXPORT_SENSITIVE,
  ],
} as const;
